File: Two Pointers/Problem4 - 1679.cpp
#include <bits/stdc++.h>
using namespace std;

/* brute force appraoch - for each element find k - ele and remove */

/* optimal appraoch - greedy + two pointer */
int maxOperations(vector<int>& nums, int k) {
    int n = nums.size();
    sort(nums.begin(), nums.end());

    int left = 0, right = n - 1;
    int ope = 0;
    while (left < right) {
        if (nums[left] + nums[right] == k) {
            ope++;
            left++, right--;
        }
        else if (nums[left] + nums[right] < k) left++;
        else right--;
    }
    return ope;
    /*  TC-O(nLogn + n) SC-O(1) */
}

File: Two Pointers/Problem4 - 1679.java
import java.util.Arrays;

public class Main {
    /* Optimal approach - greedy + two pointer */
    public static int maxOperations(int[] nums, int k) {
        Arrays.sort(nums);

        int left = 0, right = nums.length - 1;
        int ope = 0;
        while (left < right) {
            if (nums[left] + nums[right] == k) {
                ope++;
                left++;
                right--;
            } else if (nums[left] + nums[right] < k) {
                left++;
            } else {
                right--;
            }
        }
        return ope;
        /*  TC-O(nLogn + n) SC-O(1) */
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4, 5};
        int k = 5;
        System.out.println(maxOperations(nums, k)); // Output: 2
    }
}
